src/analysis/normed_space/normed_group_quotient.lean

```diff
@@ -0,0 +1,489 @@
+      apply is_open_Union,
+      rintros ⟨s, s_in⟩,
+      exact (continuous_add_right s).is_open_preimage _ is_open_ball },
+    { exact ⟨(0 : M), mem_ball_self ε_pos, (mk' S).map_zero⟩ } },
+end⟩
+
+/-- The seminormed group structure on the quotient by an additive subgroup. -/
+noncomputable
+instance add_subgroup.semi_normed_group_quotient (S : add_subgroup M) :
+  semi_normed_group (quotient S) :=
+{ dist               := λ x y, ∥x - y∥,
+  dist_self          := λ x, by simp only [norm_mk_zero, sub_self],
+  dist_comm          := quotient_norm_sub_rev,
+  dist_triangle      := λ x y z,
+  begin
+    unfold dist,
+    have : x - z = (x - y) + (y - z) := by abel,
+    rw this,
+    exact quotient_norm_add_le S (x - y) (y - z)
+  end,
+  dist_eq := λ x y, rfl,
+  to_uniform_space   := topological_add_group.to_uniform_space (quotient S),
+  uniformity_dist    :=
+  begin
+    rw uniformity_eq_comap_nhds_zero',
+    have := (quotient_nhd_basis S).comap (λ (p : quotient S × quotient S), p.2 - p.1),
+    apply this.eq_of_same_basis,
+    have : ∀ ε : ℝ, (λ (p : quotient S × quotient S), p.snd - p.fst) ⁻¹' {x | ∥x∥ < ε} =
+      {p : quotient S × quotient S | ∥p.fst - p.snd∥ < ε},
+    { intro ε,
+      ext x,
+      dsimp,
+      rw quotient_norm_sub_rev },
+    rw funext this,
+    refine filter.has_basis_binfi_principal _ set.nonempty_Ioi,
+    rintros ε (ε_pos : 0 < ε) η (η_pos : 0 < η),
+    refine ⟨min ε η, lt_min ε_pos η_pos, _, _⟩,
+    { suffices : ∀ (a b : quotient S), ∥a - b∥ < ε → ∥a - b∥ < η → ∥a - b∥ < ε, by simpa,
+      exact λ a b h h', h },
+    { simp }
+  end }
+
+-- This is a sanity check left here on purpose to ensure that potential refactors won't destroy
+-- this important property.
+example (S : add_subgroup M) : (quotient.topological_space : topological_space $ quotient S) =
+S.semi_normed_group_quotient.to_uniform_space.to_topological_space :=
+rfl
+
+/-- The quotient in the category of normed groups. -/
+noncomputable
+instance add_subgroup.normed_group_quotient (S : add_subgroup M) [hS : is_closed (S : set M)] :
+  normed_group (quotient S) :=
+{ eq_of_dist_eq_zero :=
+  begin
+    rintros ⟨m⟩ ⟨m'⟩ (h : ∥mk' S m - mk' S m'∥ = 0),
+    erw [← (mk' S).map_sub, quotient_norm_eq_zero_iff, hS.closure_eq,
+         ← quotient_add_group.eq_iff_sub_mem] at h,
+    exact h
+  end,
+  .. add_subgroup.semi_normed_group_quotient S }
+
+-- This is a sanity check left here on purpose to ensure that potential refactors won't destroy
+-- this important property.
+example (S : add_subgroup M) [is_closed (S : set M)] :
+  S.semi_normed_group_quotient = normed_group.to_semi_normed_group := rfl
+
+
+namespace add_subgroup
+
+open normed_group_hom
+
+/-- The morphism from a seminormed group to the quotient by a subgroup. -/
+noncomputable
+def normed_mk (S : add_subgroup M) : normed_group_hom M (quotient S) :=
+{ bound' := ⟨1, λ m, by simpa [one_mul] using quotient_norm_mk_le  _ m⟩,
+  ..quotient_add_group.mk' S }
```

Comment:
I don't know what convention is but I see that throughout the rest of the file you leave a space here.
```suggestion
  .. quotient_add_group.mk' S }
```